gapis/trace/android/profile/groups.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (C) 2022 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package profile
+
+import (
+	"sort"
+
+	"github.com/google/gapid/core/data/slice"
+	"github.com/google/gapid/gapis/api/sync"
+	"github.com/google/gapid/gapis/service"
+	"github.com/google/gapid/gapis/service/path"
+)
+
+type groupTreeNode struct {
+	id   int32
+	name string
+	link sync.SubCmdRange
+
+	children []groupTreeNode // sorted by child.link
+}
+
+type GroupTree struct {
+	nextID int32
+	groupTreeNode
+}
+
+func NewGroupTree() *GroupTree {
+	return &GroupTree{
+		nextID: 1,
+		groupTreeNode: groupTreeNode{
+			id:   0,
+			name: "root",
+		},
+	}
+}
+
+// CreateOrGetGroup finds or creates the group for the given command range and returns its id.
+// TODO: this function makes some assumptions about command/sub command IDs:
+// 1. we only get groups for command buffers, renderpasses and draw calls.
+// 2. no overlaps.
+// 3. the sub command ids are [cmdId, submission, cmdbuff, cmd].
+// All these assumptions currently hold and are also made in other parts of the
+// code in some way. The assumptions will need to be codified as part of the
+// command/sub-command refactor that is already planned.
+func (t *GroupTree) CreateOrGetGroup(name string, link sync.SubCmdRange) int32 {
```

Comment:
getOrCreateX() is the usual idiom here, rather than createOrGetX()

Reply:
Done.